Fix build, update base images and software, update readme

SuperSandro2000 opened the pull request:
This branch kinda grow a bit so if you want me to split something out just let me know and I split it in a separate PR or drop it entirely.

For individual information please take a look at the extended commit message as I tried to put some statistics in their.

Also feel free to question me about all the changes. Most of them have a good reason behind them but I not always documented them or they are just to fulfill my ocd.

Also if you need a person that helps maintain the docker image for musicbrainz I kinda want to apply for that as the image was quite out of date and I think some help would be appreciated.

yvanzo reviewed (changes requested):
Thanks for the proposed changes, however most of these should apply to the current development branch [`mbvm-38-dev`](https://github.com/metabrainz/musicbrainz-docker/tree/mbvm-38-dev) instead of the [`master`](https://github.com/metabrainz/musicbrainz-docker/tree/master) branch which currently is in maintenance mode.

Indeed, the `master` branch is deployed and ran by people using the current [MusicBrainz Slave Server Virtual Machine (VirtualBox)](https://musicbrainz.org/doc/MusicBrainz_Server/Setup#MusicBrainz_Server_virtual_machine). It features the old Lucene-based search server, as opposed to the new Solr-based search server supported by the `mbvm-38-dev` branch.

Thus the only patches going through the `master` branch are mostly about updating MusicBrainz Server and are actually backported from the `mbvm-38-dev` branch.

More insight about the ongoing roadmap:
1. Currently implementing MBS-10546
2. Then update `mbvm-38-dev` with that feature
3. Merge `mbvm-38-dev` branch into `master` and publish a new OVA after MBVM-38.

Please consider rebasing this pull request on `mbvm-38-dev`.

SuperSandro2000 committed:
Delete apt lists

This saves 17MB.

docker images | rg ^musicbrainz
musicbrainz-postgres-clear-apt-list                 latest                             8a1e219a6410        5 seconds ago       575MB
musicbrainz-postgres-before                         latest                             0db38dd2367a        5 minutes ago       592MB

SuperSandro2000 committed:
Effectively delete git data in all layers

If we delete the git data in a later RUN it is already saved in the before RUN.
With this change we only save 2MB but otherwise the delete at the end is not that effective.

docker images | rg ^musicbrainz
musicbrainz-postgres-combine-runs                   latest                             8ed6fa10387f        7 seconds ago       573MB
musicbrainz-postgres-clear-apt-list                 latest                             8a1e219a6410        5 minutes ago       575MB

SuperSandro2000 committed:
Add --no-install-recommends to apt

This saves 27MB.

docker images | rg ^musicbrainz
musicbrainz-postgres-no-install-recommends          latest                             58ea91bb6e33        8 seconds ago       546MB
musicbrainz-postgres-combine-runs                   latest                             8ed6fa10387f        4 minutes ago       573MB

SuperSandro2000 committed:
Remove not needed packages

This saves an additional 37MB.

docker images | rg ^musicbrainz
musicbrainz-postgres-cleanup-packages               latest                             c471150e1c88        6 seconds ago       509MB
musicbrainz-postgres-no-install-recommends          latest                             58ea91bb6e33        4 minutes ago       546MB

SuperSandro2000 committed:
Combine COPY

SuperSandro2000 committed:
Combine runs and git clone

This is mostly aesthetic and saves only 1MB.

docker images | rg ^musicbrainz-sir
musicbrainz-sir-combined-runs                       latest                             514e58687a90        4 seconds ago        730MB
musicbrainz-sir-before                              latest                             f6b9109bc41a        About a minute ago   731MB

SuperSandro2000 committed:
Cleanup not needed packages, format

This saves 133MB.

docker images | rg ^musicbrainz-sir
musicbrainz-sir-clean-packages                      latest                             9d15cb1f6ade        11 seconds ago      597MB
musicbrainz-sir-combined-runs                       latest                             514e58687a90        5 minutes ago       730M

SuperSandro2000 committed:
Combine two runs

SuperSandro2000 committed:
Put solr straight into docker-compose

SuperSandro2000 committed:
Update dockerize to 0.6.1

Changelog:
0.6.1
4c86ed7 Fix 100% cpu util when tailing file
0400d2c Fix #100 allowing to tail named pipes
85a884e add build for i386 architecture

0.6.0
492e1c8 Add file wait option
b16c4dd Don't trap sigquit so that a stack trace can be forced
14524ea Update hpcloud/tail dep

SuperSandro2000 committed:
Optimize git clone, move version to ARG

SuperSandro2000 committed:
Clean apt-get lists at the right location, remove random rm

SuperSandro2000 committed:
Remove random apt install

SuperSandro2000 committed:
Move workdir up to save some cd's

SuperSandro2000 committed:
Move node version to ARG

SuperSandro2000 committed:
Update node to 10.19.0

Musicbrainz server uses v10 since a longer time and v8 is no longer supported.

SuperSandro2000 committed:
Remove yarn and cpanm cache

This saves about 500MB image size.

docker images | rg ^musicbrainz-server
musicbrainz-server-clean-caches                     latest                             d03e8cbe82af        30 minutes ago      1.49GB
musicbrainz-server-after                            latest                             4286825af623        3 hours ago         1.92GB

SuperSandro2000 committed:
Move wait after resource compilation

This speeds up the container startup and we don't waste time waiting for the db, redis, mq to come up
and then waiting for the resources to compile.

SuperSandro2000 committed:
Fix entrypoint, the script is located in path not /

SuperSandro2000 committed:
Comment for what certain libs are used, remove no longer needed ones

This should make it easier in the future to keep track why we
installed something and why we still need it.

yvanzo commented on docker-compose.yml:
```diff
@@ -77,7 +77,7 @@ services:
 
   redis:
-    image: redis:3-alpine
+    image: redis:5-alpine
```
This is a bump of two major versions, but [Redis release cycle](https://redis.io/topics/releases) specifies they _try very hard to make the Redis API mostly backward compatible_ and _upgrading to newer versions is usually trivial_. 4.0 (blog post)[https://redislabs.com/blog/redis-4-0-0-released/]/[ml post](https://groups.google.com/d/msg/redis-db/5Kh3viziYGQ/58TKLwX0AAAJ) and 5.0 [blog post](https://redislabs.com/blog/redis-5-0-is-here/)/[ml post](https://groups.google.com/d/msg/redis-db/l0OXDAlwosU/JyYfeQpyBwAJ) doesn’t any breaking change either. So it is probably fine but Redis 3 is still in production at the moment, see https://github.com/metabrainz/docker-redis.

That would be nice to:
1. Have more detailed commit message, e.g. by including the above rationale;
2. Allow setting Redis version through a Docker environment variable for tests (before changing the default), e.g.

```suggestion
    image: redis:${REDIS_VERSION:-3}-alpine
```

(untested suggestion)

yvanzo commented on build/postgres/Dockerfile:
```diff
@@ -17,6 +17,7 @@ RUN apt-get update && \
         libpq-dev \
         libxml2-dev \
         pkg-config \
+        postgresql-client-${POSTGRES_VERSION} \
```
When does the error mentioned in 59433b315bcddbc2c2cbcb142b6a3a48c619022e happen? I cannot reproduce it.

SuperSandro2000 replied:
When I tried building the ``9.5`` postgres image ``9.5.20`` was tagged as ``9.5``:
````
/tmp/musicbrainz-docker/build/postgres $ docker pull postgres:9.5.20
9.5: Pulling from library/postgres
619014d83c02: Already exists
7ec0fe6664f6: Already exists
9ca7ba8f7764: Already exists
9e1155d037e2: Already exists
febcfb7f8870: Already exists
8c78c79412b5: Already exists
5a35744405c5: Already exists
27717922e067: Already exists
48127670a602: Pull complete
6aedb706f269: Pull complete
158b65b6f8f1: Pull complete
7e3410bd4805: Pull complete
2c187ebaa355: Pull complete
a93e50cc434a: Pull complete
Digest: sha256:40cf77833a45615aa06e5073669f7083f1b1cdd350bae1724425e3d8333389dc
Status: Downloaded newer image for postgres:9.5
docker.io/library/postgres:9.5

/tmp/musicbrainz-docker/build/postgres $ git br
  master
* mbvm-38-dev

/tmp/musicbrainz-docker/build/postgres $ docker build .
[+] Building 12.2s (9/15)
 => [internal] load build definition from Dockerfile                                                                                                                                                     0.2s
 => => transferring dockerfile: 1.17kB                                                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                                                        0.1s
 => => transferring context: 2B                                                                                                                                                                          0.0s
 => [internal] load metadata for docker.io/library/postgres:9.5.20                                                                                                                                       0.0s
 => [internal] load build context                                                                                                                                                                        0.1s
 => => transferring context: 35B                                                                                                                                                                         0.0s
 => [1/12] FROM docker.io/library/postgres:9.5.20                                                                                                                                                        0.0s
 => CACHED [2/12] RUN apt-get update &&     apt-get -y -q install         build-essential         git-core         libdb-dev         libexpat1-dev         libicu-dev         libpq-dev         libxml2  0.0s
 => [3/12] RUN git clone https://github.com/metabrainz/postgresql-musicbrainz-unaccent.git /tmp/postgresql-musicbrainz-unaccent &&     git clone https://github.com/metabrainz/postgresql-musicbrainz-  10.4s
 => [4/12] WORKDIR /tmp/postgresql-musicbrainz-unaccent                                                                                                                                                  0.1s
 => ERROR [5/12] RUN make && make install                                                                                                                                                                1.4s
------
 > [5/12] RUN make && make install:
#8 1.258 Makefile:25: /usr/lib/postgresql/12/lib/pgxs/src/makefiles/pgxs.mk: No such file or directory
#8 1.259 make: *** No rule to make target '/usr/lib/postgresql/12/lib/pgxs/src/makefiles/pgxs.mk'.  Stop.
````

But when I build specifically against ``9.5.21`` the bug does not happen.
I suggest we just track postgres ``9.5.21`` for now. 

Edit: https://github.com/docker-library/postgres/issues/677 and https://github.com/docker-library/postgres/issues/678 on the postgres library bugtracker.

yvanzo replied:
Thanks, so this is a new bug due to upstream. I can reproduce it after clearing caches, but pulling 9.5.21 doesn’t work for me either.

SuperSandro2000 replied:
So what do we do about this? It seems that the container is partially pg 12 and not 9.5 which is kinda confusing and not good at all.

yvanzo replied:
Issue was installed version of `libpq5` = 12. I pushed https://github.com/metabrainz/musicbrainz-docker/commit/af484dfc830563a917d2782b11f35451ed2e23af fix to [mbvm-38-dev](https://github.com/metabrainz/musicbrainz-docker/tree/mbvm-38-dev).

```diff
@@ -8,13 +8,14 @@ ARG POSTGRES_VERSION
 
 # Update the Ubuntu and PostgreSQL repository indexes
 RUN apt-get update && \
-    apt-get -y -q install \
+    apt-get -y -q install --allow-downgrades \
         build-essential \
         git-core \
         libdb-dev \
         libexpat1-dev \
         libicu-dev \
-        libpq-dev \
+        libpq5=${POSTGRES_VERSION}.* \
+        libpq-dev=${POSTGRES_VERSION}.* \
         libxml2-dev \
         pkg-config \
         postgresql-server-dev-${POSTGRES_VERSION}
```

SuperSandro2000 replied:
I rebased onto https://github.com/metabrainz/musicbrainz-docker/commit/af484dfc830563a917d2782b11f35451ed2e23af . I hope I did no mistake. Otherwise I will just add a fix commit.

yvanzo commented on build/rabbitmq/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM rabbitmq:3.6.5-management
+FROM rabbitmq:3.8.2-management
```
[RabbitMQ 3.7](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.7.0) has breaking changes. [RabbitMQ 3.8](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.8.0) is compatible with 3.7-compatible client libraries.

Same as for Redis, I would suggest to add more details in commit message and to allow updating through Docker environment rather than updating without prior tests.

SuperSandro2000 replied:
Would it be save to update to rabbit 3.6.16-management?

TODO Allow changing trough ENVs.

yvanzo replied:
> Would it be save to update to rabbit 3.6.16-management?

Yes.

yvanzo commented on docker-compose.yml:
```diff
@@ -61,7 +61,7 @@ services:
 
   search:
-    build: build/solr
+    image: metabrainz/mb-solr:3.1.1
```
Does fece8bb3f7f720d5698446600c88ede64b2b5446 still allow to use a local working copy of mb-solr through compose override?
If not, please drop this commit.

SuperSandro2000 replied:
You mean [this](https://github.com/metabrainz/musicbrainz-docker/blob/mbvm-38-dev/compose/publishing-all-ports.yml#L9) section? Should be. Right now we just re-tag the downloaded image.

yvanzo replied:
Something like [`compose/musicbrainz-dev.yml`](https://github.com/yvanzo/musicbrainz-docker/blob/mbs-dev/compose/musicbrainz-dev.yml) but applied to `mb-solr`.

SuperSandro2000 replied:
That works except the build part cause the Dockerfile is no longer. You would need to add one again.

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -1,7 +1,7 @@
 LABEL maintainer=jeffsturgis@gmail.com
 
-ENV DOCKERIZE_VERSION v0.5.0
+ENV DOCKERIZE_VERSION v0.6.1
```
LGTM given [v0.6.0](https://github.com/jwilder/dockerize/releases/tag/v0.6.0) and [v0.6.1](https://github.com/jwilder/dockerize/releases/tag/v0.6.1) release notes. Would be nice to add details to da8f5a0b61635c6235ed655e395ec376bab956e8 commit message.

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -28,9 +28,8 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
-    cd musicbrainz-server && \
-    git checkout v-2020-01-20
+ARG MUSICBRAINZ_VERSION=v-2020-01-20
+RUN git clone --depth=1 --branch $MUSICBRAINZ_VERSION https://github.com/metabrainz/musicbrainz-server.git musicbrainz-server
```
I would suggest to make it more specific: `MUSICBRAINZ_SERVER_VERSION`.

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -66,7 +66,6 @@ RUN cd /musicbrainz-server/ && \
 
 COPY DBDefs.pm /musicbrainz-server/lib/
 COPY scripts/* /
-RUN apt-get install -y wget
```
`wget` is required to run [`createdb.sh`](https://github.com/metabrainz/musicbrainz-docker/blob/b3f6e896039e6a1057d4c608b108946578e775c7/build/musicbrainz/scripts/createdb.sh) later on.

SuperSandro2000 replied:
``wget`` is already installed [https://github.com/metabrainz/base-image/blob/master/Dockerfile#L6] and used [here](https://github.com/metabrainz/musicbrainz-docker/blob/b3f6e896039e6a1057d4c608b108946578e775c7/build/musicbrainz/Dockerfile#L5)

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -34,7 +34,7 @@ RUN git clone --depth=1 --branch $MUSICBRAINZ_VERSION https://github.com/metabra
 WORKDIR /musicbrainz-server
 
-ARG NODE_VERSION=8.11.3
+ARG NODE_VERSION=8.17.0
```
Oops, v8 is no longer maintained according to https://nodejs.org/en/about/releases/ and MusicBrainz Server switched to v10 a long time ago. Let’s update to [v10.19.0](https://nodejs.org/en/blog/release/v10.19.0/) directly.

yvanzo commented on build/musicbrainz/scripts/start.sh:
```diff
@@ -15,5 +15,5 @@ then
 fi
 
-/start_mb_renderer.pl
+dockerize -wait tcp://db:5432 -timeout 60s -wait tcp://mq:5672 -timeout 60s -wait tcp://redis:6379 -timeout 60s /start_mb_renderer.pl
```
What happens with ef8d1a9063a48fa26ed43f84d78c30b6bb822706 if replication cron starts before `db` is available?

SuperSandro2000 replied:
Then move it to https://github.com/metabrainz/musicbrainz-docker/blob/ef8d1a9063a48fa26ed43f84d78c30b6bb822706/build/musicbrainz/scripts/start.sh#L11 ? I kinda want a kick start with the resource compiling that the rest is up and we don't waste time waiting.

yvanzo replied:
Sounds more reliable.

yvanzo commented on build/rabbitmq/scripts/docker-entrypoint-wrapper.sh:
```diff
@@ -17,7 +17,7 @@ if [ ! -e $LOCK_FILE ]; then
 if [ $# -eq 0 ]; then
-  exec /docker-entrypoint.sh rabbitmq-server
+  exec docker-entrypoint.sh rabbitmq-server
 else
-  exec /docker-entrypoint.sh $@
+  exec docker-entrypoint.sh $@
```
Not sure what 255d22ca41a254d4bc95488052f652a72c6df758 is addressing since current working directory is `/`.

SuperSandro2000 replied:
This did not work for me. The container couldn't find the script. It seems that it does not look into the ``$PATH`` env when the ``/`` is provided.

yvanzo replied:
Is `/set-config.sh` correctly found then?

SuperSandro2000 replied:
Yes, cause we copied this into the root.

````
 docker build . -t test
[sudo] password for hotpi:
[+] Building 23.8s (7/7) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                     0.1s
 => => transferring dockerfile: 187B                                                                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                                                        0.1s
 => => transferring context: 2B                                                                                                                                                                          0.0s
 => [internal] load metadata for docker.io/library/rabbitmq:3.6.16-management                                                                                                                            2.1s
 => [internal] load build context                                                                                                                                                                        0.1s
 => => transferring context: 1.28kB                                                                                                                                                                      0.0s
 => [1/2] FROM docker.io/library/rabbitmq:3.6.16-management@sha256:05bd722c6b0c747ae7f854baa9e9ceb9b47da75e39f93ad6a1db74f788b0f828                                                                     20.2s
 => => resolve docker.io/library/rabbitmq:3.6.16-management@sha256:05bd722c6b0c747ae7f854baa9e9ceb9b47da75e39f93ad6a1db74f788b0f828                                                                      0.0s
 => => sha256:9df8e9792ce67836bd5fb07d0d47e75314d0b5586e15d2a7fce0361e3577f86d 9.72kB / 9.72kB                                                                                                           0.0s
 => => sha256:02fe1bd1a85c93837c66dbd7406515fcd5b8a683d0642581e8f3a54eb35ef173 4.50MB / 4.50MB                                                                                                           3.0s
 => => sha256:f17d81b4b692f7e0d6c1176c86b81d9f2cb5ac5349703adca51c61debcfe413c 22.49MB / 22.49MB                                                                                                        11.0s
 => => sha256:66c15a50f4da989e3d3c86925f82bec705c5c25346b92aad555a7b0da8efe7b8 4.07kB / 4.07kB                                                                                                           0.7s
 => => sha256:05bd722c6b0c747ae7f854baa9e9ceb9b47da75e39f93ad6a1db74f788b0f828 2.37kB / 2.37kB                                                                                                           0.0s
 => => sha256:eada5538ca4668b81e55cb31a2eb256d9c752775e99cb256efcf1db06107d98a 3.04kB / 3.04kB                                                                                                           0.0s
 => => sha256:771c4c62018c862476441914cb38dbf2c3fa9876effa266d27fc831ade13ff8e 951.87kB / 951.87kB                                                                                                       1.7s
 => => sha256:0ae9c44cefc970d06e67e34323059f924232f944d238090c834061cd9114c6b0 27.71MB / 27.71MB                                                                                                        13.5s
 => => sha256:638d18f9679b9ccd53ce56a192a94f7a3c8ed00bcc530da0e400db720fa1cd94 7.31MB / 7.31MB                                                                                                           7.3s
 => => sha256:80e59581508db919ce2b27b4df6c50514e9c6caa470e3761b17874d84d49ffeb 2.34kB / 2.34kB                                                                                                           7.6s
 => => sha256:065f8960e5370afaafc1a02bd11e8c9d35013110701ea8ed2488ecb9c3d25d1b 148B / 148B                                                                                                               7.9s
 => => sha256:91bf444c375323df1b1c7c6e6de873ccedd7cd25acad2fb9acd413f148966104 126B / 126B                                                                                                               8.2s
 => => sha256:6b34701157a4e042113a66cf8985972f74a7e8ad93521e81e43a481aabf5b788 4.28kB / 4.28kB                                                                                                           8.6s
 => => sha256:fafd1bae2225fc6ba40d093417aa97340ade378c17641c50bd5d9f260ab235a8 121B / 121B                                                                                                               9.0s
 => => sha256:2a87cf263e6d4a15b9623183a76e20fb8d906e39a82274205207d6136f7e195d 191B / 191B                                                                                                               9.3s
 => => sha256:d2806ea66b21c7b968494b790fed1fe7f8a56fbf2000ed9d973acb74e1754e40 7.64MB / 7.64MB                                                                                                          12.9s
 => => extracting sha256:f17d81b4b692f7e0d6c1176c86b81d9f2cb5ac5349703adca51c61debcfe413c                                                                                                                2.5s
 => => extracting sha256:02fe1bd1a85c93837c66dbd7406515fcd5b8a683d0642581e8f3a54eb35ef173                                                                                                                0.5s
 => => extracting sha256:66c15a50f4da989e3d3c86925f82bec705c5c25346b92aad555a7b0da8efe7b8                                                                                                                0.2s
 => => extracting sha256:771c4c62018c862476441914cb38dbf2c3fa9876effa266d27fc831ade13ff8e                                                                                                                0.2s
 => => extracting sha256:0ae9c44cefc970d06e67e34323059f924232f944d238090c834061cd9114c6b0                                                                                                                1.5s
 => => extracting sha256:638d18f9679b9ccd53ce56a192a94f7a3c8ed00bcc530da0e400db720fa1cd94                                                                                                                0.4s
 => => extracting sha256:80e59581508db919ce2b27b4df6c50514e9c6caa470e3761b17874d84d49ffeb                                                                                                                0.0s
 => => extracting sha256:065f8960e5370afaafc1a02bd11e8c9d35013110701ea8ed2488ecb9c3d25d1b                                                                                                                0.0s
 => => extracting sha256:91bf444c375323df1b1c7c6e6de873ccedd7cd25acad2fb9acd413f148966104                                                                                                                0.0s
 => => extracting sha256:6b34701157a4e042113a66cf8985972f74a7e8ad93521e81e43a481aabf5b788                                                                                                                0.0s
 => => extracting sha256:fafd1bae2225fc6ba40d093417aa97340ade378c17641c50bd5d9f260ab235a8                                                                                                                0.0s
 => => extracting sha256:2a87cf263e6d4a15b9623183a76e20fb8d906e39a82274205207d6136f7e195d                                                                                                                0.0s
 => => extracting sha256:d2806ea66b21c7b968494b790fed1fe7f8a56fbf2000ed9d973acb74e1754e40                                                                                                                0.9s
 => [2/2] COPY [ scripts/*, / ]                                                                                                                                                                          1.1s
 => exporting to image                                                                                                                                                                                   0.2s
 => => exporting layers                                                                                                                                                                                  0.1s
 => => writing image sha256:68b7cddc8cdf9f6f8ce42f90d813cca4e40bdf53ca9e234c9967e871c6470127                                                                                                             0.0s
 => => naming to docker.io/library/test                                                                                                                                                                  0.0s

docker run test /set-config.sh

/set-config.sh: Waiting for rabbitmq-server ...

/set-config.sh: Waiting for rabbitmq-server ...
^C^C^C^C
~/src/musicbrainz-docker/build/rabbitmq docker run test ls /
bin  boot  dev  docker-entrypoint-wrapper.sh  docker-entrypoint.sh  etc  home  lib  lib64  media  mnt  opt  plugins  proc  root  run  sbin  set-config.once  set-config.sh  srv  sys  tmp  usr  var
````

yvanzo commented on build/postgres/Dockerfile:
```diff
@@ -7,30 +7,32 @@ ARG DEBIAN_FRONTEND="noninteractive"
-RUN git clone https://github.com/metabrainz/postgresql-musicbrainz-unaccent.git /tmp/postgresql-musicbrainz-unaccent && \
-    git clone https://github.com/metabrainz/postgresql-musicbrainz-collate.git /tmp/postgresql-musicbrainz-collate && \
-    git clone https://github.com/omniti-labs/pg_amqp.git /tmp/pg_amqp
+# hadolint ignore=DL3003
```
Would that be helpful to have a `.hadolint.yaml` in this repository?

SuperSandro2000 replied:
Depends on if you want to use it. I probably did this out of habbit. I can remove it if you want.

yvanzo replied:
Having such comment is alright but having a `.hadolint.yaml` would probably be even better.

SuperSandro2000 replied:
I don't think we want to apply this rule globally. It just does not make sense in this case. There are also lot more hadolint warnings mostly about quoting and switching directory. Maybe we should look at some sever or ignore them.

yvanzo reviewed (changes requested):
So many useful and clean changes, thank you! A few changes are unclear _why_ in commit message.

I did not yet review changes about not installing packages `libdb-dev` `libffi-dev` `libssl-dev` `libpq-dev` `libxml2-dev` `libxslt1-dev` `pkg-config` `postgres-`_X_ and `zlib1g-dev`.

SuperSandro2000 committed:
Update rabbitmq to 3.6.16-management

SuperSandro2000 commented:
> So many useful and clean changes, thank you! A few changes are unclear _why_ in commit message.

Thanks

libpq-dev is required by postgresql-server-dev-X.X so we don't need to explicitly install it.
And for the others I guess we don't compile as much as we used to anymore.

yvanzo commented on build/musicbrainz/scripts/start.sh:
```diff
@@ -9,11 +9,12 @@ then
 fi
 
+dockerize -wait tcp://db:5432 -timeout 60s -wait tcp://mq:5672 -timeout 60s -wait tcp://redis:6379 -timeout 60s /start_mb_renderer.pl
+
 if [ -f /crons.conf -a -s /crons.conf ]
 then
   crontab /crons.conf
   cron -f &
 fi
 
-/start_mb_renderer.pl
 start_server --port=5000 -- plackup -I lib -s Starlet -E deployment --nproc 10 --pid fcgi.pid
```
That follows https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r379899209, yes about moving `dockerize` only.



 `start_server` and `cron` stuff should not be executed at all if dockerize failed for some reason.

SuperSandro2000 replied:
So it should be enough if we do ``set -e``? Or do you want to go with ``set -eou pipefail`` to also catch errors in pipes and unset variables?

yvanzo replied:
I see no pipe here, otherwise yes, `set -e -u` should do.

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -12,21 +12,19 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
-        postgresql-${POSTGRES_VERSION} \
+        postgresql-client-${POSTGRES_VERSION} \
         python-minimal \
-        zlib1g-dev \
```
These Ubuntu packages are build dependencies for some Perl modules installed below using `cpanm`:
* `libdb-dev` :arrow_forward: `DB_File` :arrow_forward: `Cache`
* `libxml2-dev` + `zlib1g-dev` :arrow_forward: `XML::LibXML`
* `pkg-config` :arrow_forward: `Unicode::ICU::Collator`

SuperSandro2000 replied:
I am just curious: Why did the image built without them?

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -12,21 +12,19 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
         memcached \
-        pkg-config \
-        postgresql-${POSTGRES_VERSION} \
+        postgresql-client-${POSTGRES_VERSION} \
```
`postgresql-9.5` did provide `pg_ctl` which is needed to run tests, `postgres-client-`_X_ doesn’t.

SuperSandro2000 replied:
I think this should be separated to the test image. The full postgres package is not that tiny and I don't think this is needed in the production image, is it?

yvanzo replied:
Indeed, I doesn’t seem to be needed elsewhere, it seems safe to install just the client instead.

yvanzo commented on build/sir/Dockerfile:
```diff
@@ -11,18 +11,14 @@ LABEL org.label-schema.schema-version="1.0.0-rc1" \
-                    ca-certificates \
-                    cron \
-                    git \
-                    libpq-dev \
```
`libpq-dev` is required at install `psycopg2` Python package.

SuperSandro2000 replied:
oh, yeah. I double checked that with ``ld`` and pip just does not throw any error if `libpq-dev` is missing.

yvanzo replied:
To clarify, installing without it does not throw any error but running without it fails.

https://www.psycopg.org/docs/install.html#install-from-source

yvanzo commented on build/sir/Dockerfile:
```diff
@@ -11,18 +11,14 @@ LABEL org.label-schema.schema-version="1.0.0-rc1" \
-                    cron \
-                    git \
-                    libpq-dev \
-                    libffi-dev \
-                    libssl-dev \
-                    libxml2-dev \
-                    libxslt1-dev && \
```
This comes from sir Dockerfile but I did not find any reason why these would be needed. Since it requires further live testing, let’s try removing these in a later PR.

SuperSandro2000 replied:
I don't see a reason why we should keep it then but I'll open a new PR with it removed. Better safe then sorry.

yvanzo replied:
`libxml2-dev` and `libxslt1-dev` are required by `lxml` which is installed with `mb-rngpy`.

Still looking for `libffi-dev` and `libssl-dev`.

SuperSandro2000 replied:
``libxslt1-dev`` has a dependency on ``libxml2-dev`` so we could drop that.
I could imagine ``libffi-dev`` is needed for a python package which compiles C code.

yvanzo replied:
> `libxslt1-dev` has a dependency on `libxml2-dev` so we could drop that.

Not sure it would improve anything but making harder to check dependencies.

>  I could imagine `libffi-dev` is needed for a python package which compiles C code.

Either to install a Python package or to use `ctypes` in [`sir/indexing.py`](https://github.com/metabrainz/sir/blob/v1.0.2/sir/indexing.py).

I could imagine `libssl-dev` might be required too as `sir` opens a lot of connections through a variety of Python packages.

SuperSandro2000 replied:
I put `libssl-dev` back in.

yvanzo replied:
Finally, `libffi-dev` seems safe to remove, see https://github.com/metabrainz/musicbrainz-docker/pull/123#pullrequestreview-371835081.

yvanzo commented on build/postgres/Dockerfile:
```diff
@@ -9,15 +9,11 @@ ARG POSTGRES_VERSION
         libicu-dev \
-        libpq-dev \
-        libxml2-dev \
-        pkg-config \
```
`pkg-config` (and probably some other packages) seem to be required to install postgres extensions.

SuperSandro2000 replied:
I thought we build the postgres extensions in [this image](https://github.com/metabrainz/musicbrainz-docker/blob/mbvm-38-dev/build/postgres/Dockerfile) and they are server side only. Is this correct?

yvanzo replied:
Yes but is your link on [_this image_](https://github.com/metabrainz/musicbrainz-docker/blob/mbvm-38-dev/build/postgres/Dockerfile) correct? Because it points to `build/postgres/Dockerfile` this conversation is all about.

SuperSandro2000 replied:
Yeah, you are right. I thought we where in the other file.

SuperSandro2000 replied:
I just noticed this very dangerous bug:
````
make: pkg-config: Command not found
cp sql/musicbrainz_collate.sql sql/musicbrainz_collate--1.0.0.sql
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC  -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o musicbrainz_collate.o musicbrainz_collate.c
make: pkg-config: Command not found
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -shared -o musicbrainz_collate.so musicbrainz_collate.o -L/usr/lib/x86_64-linux-gnu  -Wl,-z,relro -Wl,-z,now  -L/usr/lib/x86_64-linux-gnu/mit-krb5 -Wl,--as-needed
0
````
We don't fail if it can't compile. This is the reason why the image built without pkg-config and I did not look very closely into the log.

SuperSandro2000 replied:
It seems to compile without the dev libs as strangely the ``.h`` are already there.
The logs are also clean.
````
#6 [3/5] RUN git clone https://github.com/metabrainz/postgresql-musicbrainz...
#6 0.554 Cloning into '/tmp/postgresql-musicbrainz-unaccent'...
#6 1.598 Cloning into '/tmp/postgresql-musicbrainz-collate'...
#6 2.686 Cloning into '/tmp/pg_amqp'...
#6 3.881 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o musicbrainz_unaccent.o musicbrainz_unaccent.c
#6 3.946 musicbrainz_unaccent.c: In function ‘musicbrainz_dunaccentdict_lexize’:
#6 3.946 musicbrainz_unaccent.c:319:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
#6 3.946      char *old_locale = strdup(setlocale(LC_CTYPE, NULL));
#6 3.946      ^~~~
#6 4.018 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -shared -o musicbrainz_unaccent.so musicbrainz_unaccent.o -L/usr/lib/x86_64-linux-gnu  -Wl,-z,relro
 -Wl,-z,now  -L/usr/lib/x86_64-linux-gnu/mit-krb5 -Wl,--as-needed
#6 4.035 cp sql/musicbrainz_unaccent.sql sql/musicbrainz_unaccent--1.0.0.sql
#6 4.090 /bin/mkdir -p '/usr/lib/postgresql/9.5/lib'
#6 4.091 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 4.092 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 4.092 /bin/mkdir -p '/usr/share/doc/postgresql-doc-9.5/extension'
#6 4.093 /usr/bin/install -c -m 755  musicbrainz_unaccent.so '/usr/lib/postgresql/9.5/lib/musicbrainz_unaccent.so'
#6 4.098 /usr/bin/install -c -m 644 .//musicbrainz_unaccent.control '/usr/share/postgresql/9.5/extension/'
#6 4.099 /usr/bin/install -c -m 644 .//sql/uninstall_musicbrainz_unaccent.sql .//sql/musicbrainz_unaccent.sql sql/musicbrainz_unaccent--1.0.0.sql '/usr/share/postgresql/9.5/extension/'
#6 4.101 /usr/bin/install -c -m 644 .//README.musicbrainz_unaccent.md '/usr/share/doc/postgresql-doc-9.5/extension/'
#6 4.150 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC  -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FO
RTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o musicbrainz_collate.o musicbrainz_collate.c
#6 4.233 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -shared -o musicbrainz_collate.so musicbrainz_collate.o -L/usr/lib/x86_64-linux-gnu  -Wl,-z,relro -
Wl,-z,now  -L/usr/lib/x86_64-linux-gnu/mit-krb5 -Wl,--as-needed  -licui18n -licuuc -licudata
#6 4.263 cp sql/musicbrainz_collate.sql sql/musicbrainz_collate--1.0.0.sql
#6 4.306 /bin/mkdir -p '/usr/lib/postgresql/9.5/lib'
#6 4.307 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 4.309 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 4.310 /bin/mkdir -p '/usr/share/doc/postgresql-doc-9.5/extension'
#6 4.311 /usr/bin/install -c -m 755  musicbrainz_collate.so '/usr/lib/postgresql/9.5/lib/musicbrainz_collate.so'
#6 4.315 /usr/bin/install -c -m 644 .//musicbrainz_collate.control '/usr/share/postgresql/9.5/extension/'
#6 4.317 /usr/bin/install -c -m 644 .//sql/musicbrainz_collate.sql sql/musicbrainz_collate--1.0.0.sql '/usr/share/postgresql/9.5/extension/'
#6 4.318 /usr/bin/install -c -m 644 .//README.musicbrainz_collate.md '/usr/share/doc/postgresql-doc-9.5/extension/'
#6 4.389 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/pg_amqp.o src/pg_amqp.c
#6 4.478 src/pg_amqp.c: In function ‘amqp_local_phase2’:
#6 4.478 src/pg_amqp.c:99:3: warning: enumeration value ‘XACT_EVENT_PARALLEL_COMMIT’ not handled in switch [-Wswitch]
#6 4.478    switch(event) {
#6 4.478    ^~~~~~
#6 4.478 src/pg_amqp.c:99:3: warning: enumeration value ‘XACT_EVENT_PARALLEL_ABORT’ not handled in switch [-Wswitch]
#6 4.478 src/pg_amqp.c:99:3: warning: enumeration value ‘XACT_EVENT_PRE_COMMIT’ not handled in switch [-Wswitch]
#6 4.478 src/pg_amqp.c:99:3: warning: enumeration value ‘XACT_EVENT_PARALLEL_PRE_COMMIT’ not handled in switch [-Wswitch]
#6 4.478 src/pg_amqp.c:99:3: warning: enumeration value ‘XACT_EVENT_PRE_PREPARE’ not handled in switch [-Wswitch]
#6 4.478 src/pg_amqp.c: In function ‘local_amqp_get_a_bs’:
#6 4.478 src/pg_amqp.c:140:1: warning: type of ‘broker_id’ defaults to ‘int’ [-Wimplicit-int]
#6 4.478  local_amqp_get_a_bs(broker_id) {
#6 4.478  ^~~~~~~~~~~~~~~~~~~
#6 4.478 src/pg_amqp.c: In function ‘local_amqp_get_bs’:
#6 4.478 src/pg_amqp.c:152:1: warning: type of ‘broker_id’ defaults to ‘int’ [-Wimplicit-int]
#6 4.478  local_amqp_get_bs(broker_id) {
#6 4.478  ^~~~~~~~~~~~~~~~~
#6 4.479 src/pg_amqp.c: In function ‘local_amqp_disconnect’:
#6 4.479 src/pg_amqp.c:239:1: warning: type of ‘broker_id’ defaults to ‘int’ [-Wimplicit-int]
#6 4.479  local_amqp_disconnect(broker_id) {
#6 4.479  ^~~~~~~~~~~~~~~~~~~~~
#6 4.599 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_api.o src/librabbitmq/amqp_api.c
#6 4.715 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_connection.o src/librabbitmq/amqp_connection.c
#6 4.833 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_debug.o src/librabbitmq/amqp_debug.c
#6 4.855 src/librabbitmq/amqp_debug.c:44:6: warning: no previous prototype for ‘amqp_dump’ [-Wmissing-prototypes]
#6 4.855  void amqp_dump(void const *buffer, size_t len) {
#6 4.855       ^~~~~~~~~
#6 4.886 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_framing.o src/librabbitmq/amqp_framing.c
#6 5.611 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_mem.o src/librabbitmq/amqp_mem.c
#6 5.668 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_socket.o src/librabbitmq/amqp_socket.c
#6 5.788 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I. -I./ -I/usr/include/postgresql/9.5/server -I/usr/include/postgresql/internal -Wdate-time -D_FOR
TIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o src/librabbitmq/amqp_table.o src/librabbitmq/amqp_table.c
#6 5.871 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standa
rd -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -shared -o pg_amqp.so src/pg_amqp.o src/librabbitmq/amqp_api.o src/librabbitmq/amqp_connection.o sr
c/librabbitmq/amqp_debug.o src/librabbitmq/amqp_framing.o src/librabbitmq/amqp_mem.o src/librabbitmq/amqp_socket.o src/librabbitmq/amqp_table.o -L/usr/lib/x86_64-linux-gnu  -Wl,-z,relro -Wl,-z,now  -L
/usr/lib/x86_64-linux-gnu/mit-krb5 -Wl,--as-needed
#6 5.891 cat sql/tables/tables.sql sql/functions/functions.sql > sql/amqp--0.4.2.sql
#6 5.941 /bin/mkdir -p '/usr/lib/postgresql/9.5/lib'
#6 5.942 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 5.944 /bin/mkdir -p '/usr/share/postgresql/9.5/extension'
#6 5.945 /bin/mkdir -p '/usr/share/doc/postgresql-doc-9.5/extension'
#6 5.946 /usr/bin/install -c -m 755  pg_amqp.so '/usr/lib/postgresql/9.5/lib/pg_amqp.so'
#6 5.951 /usr/bin/install -c -m 644 .//amqp.control '/usr/share/postgresql/9.5/extension/'
#6 5.952 /usr/bin/install -c -m 644 .//updates/amqp--0.4.0--0.4.1.sql .//updates/amqp--0.3.0--0.4.0.sql .//sql/amqp--0.4.2.sql  '/usr/share/postgresql/9.5/extension/'
#6 5.954 /usr/bin/install -c -m 644 .//doc/amqp.md '/usr/share/doc/postgresql-doc-9.5/extension/'
#6 DONE 6.0s
````

yvanzo replied:
Issue reported in initial comment fixed in rebased commit 3c49ebaffcb2d7a2e161993660b98f83d35abc24.

Reported Makefile issue as MBS-10670.

yvanzo reviewed (changes requested):
Finally reviewed changes about installed packages.

It is sometimes unclear which packages are actually needed and why. Would comments help with?
```shell
        libucu-dev \ # build requirement to postgres collate extension
 ```

SuperSandro2000 committed:
Exit start script on error

This prevents starting the server and cron if no db and redis connections
can be made within 60 seconds or some other error occurs before that.

This also works around a bug in Perl local::lib that has been fixed in 2.000019.
When the base image is updated line 2 can be set straight to `set -eu`.

SuperSandro2000 commented:
> Finally reviewed changes about installed packages.
> 
> It is sometimes unclear which packages are actually needed and why. Would comments help with?
> 
> ```shell
>         libucu-dev \ # build requirement to postgres collate extension
> ```

Also if we don't built the extensions in this repo we don't need this lib or am I missing something?
Edit: Yeah, i did.

SuperSandro2000 commented:
I think we have now all changes reviewed except the last open discussion I linked below.

Open discussions:
https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r380704124
Things for a new PR:
https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r380699429

yvanzo commented on build/sir/Dockerfile:
```diff
@@ -32,16 +32,14 @@ ARG SIR_VERSION=1.0.2
-
-ENV POSTGRES_USER musicbrainz
-ENV POSTGRES_PASSWORD musicbrainz
+ENV POSTGRES_USER=musicbrainz
+ENV POSTGRES_PASSWORD=musicbrainz
```
`=` additions seem to change nothing but git blame?

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -45,8 +45,8 @@ RUN cd musicbrainz-server && \
-    eval $( perl -Mlocal::lib) && cpanm --installdeps --notest .
-RUN eval $( perl -Mlocal::lib) && cpanm --notest \
+    eval $( perl -Mlocal::lib) && cpanm --installdeps --notest . && \
+    eval $( perl -Mlocal::lib) && cpanm --notest \
```
` eval $( perl -Mlocal::lib)` doesn’t have to be run twice in the same shell.

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -12,20 +12,26 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
         libpq-dev \
-        libxml2-dev \
         libssl-dev \
+        # Needed for XML::LibXML and lxml which is installed with mb-rngpy
```
`mb-rngpy` (thus `lxml`) is a dependency of `sir` (indexer), not musicbrainz server.

```suggestion
        # Needed for XML::LibXML
```

yvanzo commented on build/musicbrainz/Dockerfile:
```diff
@@ -12,20 +12,26 @@ RUN curl -sSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
     apt-get install -y \
         cpanminus \
-        build-essential \
+        bzip2 \
```
Please install `xz-utils` too, as it is needed to decompress sample data.

yvanzo reviewed (changes requested):
I reviewed the PR again with your latest added changes, found minor issues only.

SuperSandro2000 commented:
I fixed the hopefully last 3 comments.

yvanzo reviewed (commented):
Thank you for having addressed these hopefully last 3 issues. :+1: 

However, `build/sir/Dockerfile` is missing libraries installation. The two following conversations were marked as resolved but are not actually resolved (any longer? maybe fixes have been lost on rebase):
* https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r380696137
* https://github.com/metabrainz/musicbrainz-docker/pull/123#discussion_r380699429

Also, you might want to squash the two latest commits 9be52f6d3a3882bacc3a0c690cca8ed15788d599 and e02341dd31ecc392e827b148b9b81c91c066968c into the commit ed92058311b8a8e1a9fd69713509c223b638c95d these actually fix.

yvanzo commented:
Hi @SuperSandro2000, please let me know if I can help you with rebasing this very promising PR or just merging it and restoring packages `build/sir/Dockerfile` is missing afterwards. Thank you!

SuperSandro2000 commented:
- [x] squashed 9be52f6 and e02341d into ed92058

> The two following conversations were marked as resolved 

I think they got marked as outdated and I just missed them after that. I hope I resolved them now.

Also there was nothing to rebase. I am not sure if this is an error on my end.

yvanzo commented on build/musicbrainz/scripts/start.sh:
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
+set -eu
 
 eval $( perl -Mlocal::lib )
```
This just breaks due to a [bug in Perl local::lib](https://rt.cpan.org/Public/Bug/Display.html?id=112711) that has been [fixed in 2.000019](https://metacpan.org/changes/distribution/local-lib#L34).
Here is a simple workaround until base image gets updated:

```suggestion
set -e

# liblocal-lib-perl < 2.000019 generates commands using unset variable
eval $( perl -Mlocal::lib )

set -u
```

SuperSandro2000 replied:
Fixed in 1c39dac2e8f33874089d4058bb96c0a3f1670967

yvanzo reviewed (changes requested):
I thoroughly tested the indexer (going through complete reinstall, extension, trigger, live reindex, adding entity, and searching for entity) to make sure that removing of `libffi-dev` was safe indeed.

But I encountered a breaking bug in last commit https://github.com/metabrainz/musicbrainz-docker/pull/123/commits/5761cb38cf26e10f8ede62277097393f13debcf8.

yvanzo commented:
Thanks!

SuperSandro2000 commented:
No problemo. Let me know if there rise up any issues or you need/want help with updating the base image or a second opinion or review.
